api_definition.go

```diff
@@ -1087,27 +1087,14 @@ func (a *APISpec) Version(r *http.Request) (*apidef.VersionInfo, []URLSpec, bool
 }
 
 type RoundRobin struct {
-	sync.Mutex
-	pos, max int
+	pos uint32
 }
 
-func (r *RoundRobin) SetMax(max int) {
-	if r.max = max; r.max < 0 {
-		r.max = 0
+func (r *RoundRobin) WithLen(len int) int {
+	if len < 1 {
+		return 0
 	}
-
-	// Can't have a new list substituted that's shorter
-	if r.pos > r.max {
-		r.pos = 0
-	}
-}
-
-func (r *RoundRobin) SetLen(len int) { r.SetMax(len - 1) }
-
-func (r *RoundRobin) Pos() int {
-	cur := r.pos
-	if r.pos++; r.pos > r.max {
-		r.pos = 0
-	}
-	return cur
+	// -1 to start at 0, not 1
+	cur := atomic.AddUint32(&r.pos, 1) - 1
```

Comment:
It's unsigned, so it will just go back to 0.

I the number of bits was small and the list was very long, this could mean that we wouldn't get to all the elements. But given that it's 32 bits and lists will rarely ever go beyond the dozens, it's fine.

For what it's worth, the existing implementation also has overflows. But with this one it's more obvious and more predictable, as the number is always unsigned and 32 bits.